sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -61,95 +173,203 @@ Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+happened). This means that transparently a node can depend on a previously
+recorded graph (sub-graph), which works by creating edges to the individual nodes
+in the old graph. Explicit memory operations without kernels, such as a memory
+copy, are still classed as nodes under this definition, as the
+{explicit-memory-ops}[SYCL 2020 specification states] that these can be seen as
+specialized kernels executing on the device.
+
+In the explicit graph building API, nodes can also represent a memory allocation/free
+operation on the device.
+
+| Edge
+| Dependency between commands as a happens-before relationship.
+
+When recording a queue to construct a graph, an edge in the SYCL graph represents
+a data dependency between two nodes. These dependencies are expressed by the user
+code through buffer accessors. There is also the partial ability to track USM
+data dependencies provided the pointers used in the graph nodes are the same.
+With the limitation that a node taking an offset USM pointer input will not be
+identified as having an edge to another node taking a pointer input to the base
+address of the same USM allocation.
+
+In the explicit graph building API, `make_edge()` is used to define the dependency
+rather than inferring them from data dependencies.
 |===
 
-== Node
+=== API Modifications
 
-Node is a class that encapsulates tasks like SYCL kernel functions or host tasks for deferred execution.
-A graph has to be created first, the structure of a graph is defined second by adding nodes and edges.
+[source, c++]
+----
+namespace sycl {
+namespace ext::oneapi::experimental {
+
+// State of a queue, returned by info::queue::state
+enum class queue_state {
+  executing,
+  recording
+};
+
+class node {};
+
+void make_edge(node sender, node receiver);
+
+// State of a graph
+enum class graph_state {
+  modifiable,
+  executable
+};
+
+// New object representing graph
+template<graph_state State = graph_state::modifiable>
+class command_graph {};
+
+template<>
+class command_graph<graph_state::modifiable> {
+public:
+  command_graph(const property_list &propList = {});
+  command_graph<graph_state::executable> finalize(const context &syclContext) const;
+
+  node add(const std::vector<node>& dep = {});
+
+  template<typename T>
+  node add(T cgf, const std::vector<node>& dep = {});
+
+  node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& dep = {});
+  node add_free(void *data, const std::vector<node>& dep = {});
+};
+
+template<>
+class command_graph<graph_state::executable> {
+public:
+    command_graph() = delete;
+    void update(const command_graph<graph_state::modifiable> &graph);
+};
+}  // namespace ext::oneapi::experimental
+
+// New methods added to the sycl::queue class
+using namespace ext::oneapi::experimental;
+class queue {
+public:
+  bool begin_recording(command_graph<graph_state::modifiable> &graph);
+  bool end_recording();
+  event submit(command_graph<graph_state::executable> graph);
+};
```

Comment:
As previously discussed we could switch this submit call to accept a handler in line with regular queue submits, that would enable us to use `depends_on` to schedule the graph with other queue submissions, like so:
```c++
auto ev = q.submit([&](handler& cgh){
  cgh.depends_on(some_other_event);
  //Not sure of the most appropriate name here
  cgh.exec_graph(g);
});
```
I think this makes sense in keeping with exisiting submission semantics, and it shouldn't present any implementation difficulties. I am keen to hear others thoughts and suggestions though!

Reply:
Agree with this, is more symmetric with the rest of the SYCL API. We could also add a queue convenience shortcut function for this. Don't have a better suggestion for naming than `exec_graph` though.

Reply:
Moved this discussion to https://github.com/reble/llvm/issues/21 and can make a standalone follow-up PR for it